src/js_native_api_v8.cc

```diff
@@ -775,6 +802,121 @@ napi_status NAPI_CDECL napi_get_last_error_info(
+  std::vector<std::string> errors_vec;
+
+  v8impl::PlatformWrapper* platform =
+      new v8impl::PlatformWrapper(argc, argv, exec_argc, exec_argv);
+  if (platform->args.size() < 1) platform->args.push_back("libnode");
+
+  int exit_code = node::InitializeNodeWithArgs(
+      &platform->args, &platform->exec_args, &errors_vec);
+
+  HANDLE_ERRORS_VECTOR(errors, errors_vec);
+
+  if (exit_code != 0) {
+    return napi_generic_failure;
+  }
+
+  if (thread_pool_size <= 0)
+    thread_pool_size = node::per_process::cli_options->v8_thread_pool_size;
+
+  platform->platform = node::MultiIsolatePlatform::Create(thread_pool_size);
+  v8::V8::InitializePlatform(platform->platform.get());
+  v8::V8::Initialize();
+  *result = reinterpret_cast<napi_platform>(platform);
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL napi_destroy_platform(napi_platform platform) {
+  auto wrapper = reinterpret_cast<v8impl::PlatformWrapper*>(platform);
+  v8::V8::Dispose();
+  v8::V8::DisposePlatform();
+
+  // The node::CommonEnvironmentSetup::Create uniq_ptr is destroyed here
+  delete wrapper;
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL napi_create_environment(napi_platform platform,
+                                               char*** errors,
+                                               const char* main_script,
+                                               napi_env* result) {
+  auto wrapper = reinterpret_cast<v8impl::PlatformWrapper*>(platform);
+  std::vector<std::string> errors_vec;
+
+  auto instance_data = new v8impl::EnvironmentInstanceData(
+      node::CommonEnvironmentSetup::Create(wrapper->platform.get(),
+                                           &errors_vec,
+                                           wrapper->args,
+                                           wrapper->exec_args));
+
+  if (instance_data->setup() == nullptr) {
+    HANDLE_ERRORS_VECTOR(errors, errors_vec);
+    return napi_generic_failure;
+  }
+
+  v8::HandleScope handle_scope(instance_data->setup()->isolate());
+  v8::Local<v8::Context> context = instance_data->setup()->context();
+  v8::Context::Scope context_scope(context);
+
+  v8::MaybeLocal<v8::Value> loadenv_ret =
+      node::LoadEnvironment(instance_data->setup()->env(), main_script);
+
+  auto env__ = new node_napi_env__(context, wrapper->args[1]);
+  env__->instance_data = reinterpret_cast<void*>(instance_data);
+  env__->node_env()->AddCleanupHook(
+      [](void* arg) { static_cast<napi_env>(arg)->Unref(); },
+      static_cast<void*>(env__));
+  *result = env__;
+
+  if (loadenv_ret.IsEmpty()) return napi_pending_exception;
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL napi_destroy_environment(napi_env env, int* exit_code) {
+  CHECK_ARG(env, env);
+  node_napi_env node_env = reinterpret_cast<node_napi_env>(env);
+
+  {
+    int r = node::SpinEventLoop(node_env->node_env()).FromMaybe(1);
+    if (exit_code != nullptr) *exit_code = r;
+    node::Stop(node_env->node_env());
+  }
```

Comment:
It’s going to be heavily counterintuitive that this is part of a destruction operation.

Reply:
Once you have drained the event loop, you have no other course of action then to destroy the environment - this is the reason I did it this way. There is no point in having a separate API call, if you will always have to call both of them one after the other. The real question is can we (and should we) keep an environment when its event loops is empty? This would allow to call async functions in a persistent environment but it will introduce a significant difference between the stand-alone Node.js and a embedded Node.js.

Reply:
You could want to take an action in response to the event loop being drained, though? 

> The real question is can we (and should we) keep an environment when its event loops is empty? 

I think the answer to both questions here is yes.

Reply:
Ok, but this would require splitting `node::SpinEventLoop` into iteration and cleanup and it is something that is not possible with the current embedders' API.

Reply:
… but it is? The docs for `SpinEventLoop` say which steps it takes, and all of those are public API. `SpinEventLoop()` is a helper, not a building block, and exists only for convenience and stronger API stability for those for whom it is sufficient.

Reply:
Ok, I added `napi_run_environment` that spins up the event loop without destroying the environment when it is emptied and a unit test with an async callback.